File: 프로그래머스/2/42883. 큰 수 만들기/큰 수 만들기.py
def solution(number, k):

    stack = []
    for v in number:
        while stack and stack[-1] < v and k>0:
            stack.pop()
            k -= 1
        stack.append(v)

    if k > 0:
        stack = stack[:-k]
    return ''.join(stack)

File: 프로그래머스/2/42883. 큰 수 만들기/큰 수 만들기.java
import java.util.*;
class Solution {
    public static String solution(String number, int k) {
        List<Integer> stack = new ArrayList<>();

        for(int i=0; i<number.length(); i++){
            int a =number.charAt(i) - '0';

            while(stack.size()>0 && stack.get(stack.size()-1) < a && k>0){
                stack.remove(stack.size() - 1);
                k-=1;
            }
            stack.add(a);
        }
        
        while (k > 0) {
            stack.remove(stack.size() - 1);
            k--;
        }

        StringBuilder sb = new StringBuilder();
        for (int a : stack) {
            sb.append(a);
        }

        return sb.toString();
    }
}